File: Merge_Intervals/merge_intervals.js
/**
 * @param {number[][]} intervals - The array of intervals
 * @returns {number[][]} returns the summary merge intervals
 */
var merge = function (intervals) {
	const result = [];

	if (intervals.length === 0) {
		return result;
	}

	intervals.sort((a, b) => a[0] - b[0]);

	result.push(intervals[0]);

	for (let i = 1; i < intervals.length; i++) {
		let current = intervals[i];
		let resultBound = result[result.length - 1];

		if (current[0] <= resultBound[1]) {
			resultBound[1] = Math.max(resultBound[1], current[1]);
		} else {
			result.push(current);
		}
	}

	return result;
};

File: Merge_Intervals/merge_intervals.go
package mergeintervals

import "sort"

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func merge(intervals [][]int) [][]int {
	result := [][]int{}

	if len(intervals) == 0 {
		return result
	}

	sort.Slice(intervals, func(i, j int) bool { return intervals[i][0] < intervals[j][0] })

	result = append(result, intervals[0])

	for i := 1; i < len(intervals); i++ {
		current := intervals[i]
		resultBound := result[len(result)-1]

		if current[0] <= resultBound[1] {
			resultBound[1] = max(resultBound[1], current[1])
		} else {
			result = append(result, current)
		}
	}

	return result
}
